openraft/src/testing/suite.rs

```diff
@@ -0,0 +1,1511 @@
+                }])
+                .await?;
+
+            let (applied, membership) = store.last_applied_state().await?;
+            assert_eq!(Some(LogId { term: 1, index: 5 }), applied);
+            assert_eq!(
+                Some(EffectiveMembership {
+                    log_id: LogId { term: 1, index: 3 },
+                    membership: Membership::new_single(btreeset! {1,2})
+                }),
+                membership
+            );
+        }
+
+        Ok(())
+    }
+
+    pub async fn delete_logs_from(builder: &B) -> anyhow::Result<()> {
+        tracing::info!("--- delete start == stop");
+        {
+            let store = builder.build().await;
+            Self::feed_10_logs_vote_self(&store).await?;
+
+            store.delete_logs_from(1..1).await?;
+
+            let logs = store.get_log_entries(1..11).await?;
+            assert_eq!(logs.len(), 10, "expected all (10) logs to be preserved");
+        }
+
+        tracing::info!("--- delete start < stop");
+        {
+            let store = builder.build().await;
+            Self::feed_10_logs_vote_self(&store).await?;
+
+            store.delete_logs_from(..=0).await?;
+
+            store.delete_logs_from(1..4).await?;
+
+            let logs = store.try_get_log_entries(0..100).await?;
+            assert_eq!(logs.len(), 7);
+            assert_eq!(logs[0].log_id.index, 4);
+        }
+
+        tracing::info!("--- delete start < large stop");
+        {
+            let store = builder.build().await;
+            Self::feed_10_logs_vote_self(&store).await?;
+
+            store.delete_logs_from(..=0).await?;
+
+            store.delete_logs_from(1..1000).await?;
+            let logs = store.try_get_log_entries(0..).await?;
+
+            assert_eq!(logs.len(), 0);
+        }
+
+        tracing::info!("--- delete start, None");
+        {
+            let store = builder.build().await;
+            Self::feed_10_logs_vote_self(&store).await?;
+
+            store.delete_logs_from(..=0).await?;
+
+            store.delete_logs_from(1..).await?;
+            let logs = store.try_get_log_entries(0..100).await?;
+
+            assert_eq!(logs.len(), 0);
+        }
+
+        Ok(())
+    }
+
+    pub async fn append_to_log(builder: &B) -> anyhow::Result<()> {
+        let store = builder.build().await;
+        Self::feed_10_logs_vote_self(&store).await?;
+
+        store.delete_logs_from(..=0).await?;
+
+        store
+            .append_to_log(&[&Entry {
+                log_id: (2, 10).into(),
+                payload: EntryPayload::Blank,
+            }])
+            .await?;
+
+        let l = store.try_get_log_entries(0..).await?.len();
+        let last = store.try_get_log_entries(0..).await?.last().unwrap().clone();
+
+        assert_eq!(l, 10, "expected 10 entries to exist in the log");
+        assert_eq!(last.log_id, (2, 10).into(), "unexpected log id");
+        Ok(())
+    }
+
+    // pub async fn apply_single(builder: &B) -> anyhow::Result<()> {
```

Comment:
Are these comments useful?

Reply:
Yes. To provide state machine tests requires other APIs to be provided by `RaftStorage`.
These tests will be enabled when these apis are ready.

The hard part is that the state machine is quite business-related. It's hard to extract abstraction out of it for testing.